File: Java/Trees/VerticalOrderofBT.java
/**
 * Definition for binary tree
 * class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ArrayList<ArrayList<Integer>> verticalOrderTraversal(TreeNode A) {
        ArrayList<ArrayList<Integer>> res = new ArrayList<ArrayList<Integer>>();
        if(A == null)return res;

        // Get left and right levels
        int[] levels = new int[2];
        levels[0] = Integer.MAX_VALUE; // left will have negative level values
        levels[1] = Integer.MIN_VALUE; // right will have positive level values
        getExtremes(A, 0, levels);
        int tot = levels[1] - levels[0]; // levels[0] is negative
        for(int i = 0; i <= tot; i++)
            res.add(new ArrayList<Integer>());
        int start = -levels[0];
        /*
            Instead of getting extreme level values, you can just add TreeNodes in hashmap
            For more info: check hint ans;
            We will save some time by sacrificing memory
        */

        Queue<Node> queue = new LinkedList<Node>();
        queue.add(new Node(A, 0));
        while(!queue.isEmpty()){
            Node node = queue.poll();
            res.get(node.level+start).add(node.tree.val);
            if(node.tree.left != null)queue.add(new Node(node.tree.left, node.level-1));
            if(node.tree.right != null)queue.add(new Node(node.tree.right, node.level+1));
        }
        return res;
    }

    public void getExtremes(TreeNode a, int level, int[] levels){
        if(a == null)return;
        if(level < levels[0])levels[0] = level;
        if(level > levels[1])levels[1] = level;
        getExtremes(a.left, level - 1, levels);
        getExtremes(a.right, level + 1, levels);
    }

    class Node{
        TreeNode tree;
        int level;
        public Node(TreeNode n, int l){
            tree = n;
            level = l;
        }
    }
}

File: JS/Tree/VerticalOrderofBT.js
// Definition for a  binary tree node
//    function TreeNode(data){
//      this.data = data
//      this.left = null
//      this.right = null
//    }

module.exports = {
 //param root : root node of tree
 //return a array of array of integers
	verticalOrderTraversal : function(root){
        var res = [], map = {}, minLevel = 99999999;
        var queue = [{node: root, level : 0}];
        while(queue.length > 0){
            var TreeNode = queue.shift();
            var node = TreeNode.node, level = TreeNode.level;
            if(level < minLevel)minLevel = level;
            if(!map.hasOwnProperty(level))map[level] = [];
            map[level].push(node.data);

            if(node.left != null)queue.push({node: node.left, level : level - 1});
            if(node.right != null)queue.push({node: node.right, level: level + 1});
        }

        while(map.hasOwnProperty(minLevel))
            res.push(map[minLevel++]);
        return res;
	}
};
